src/uct/sm/mm/base/mm_iface.c

```diff
@@ -237,66 +251,94 @@ static inline ucs_status_t uct_mm_iface_process_recv(uct_mm_iface_t *iface,
-    read_index_elem = UCT_MM_IFACE_GET_FIFO_ELEM(iface, iface->recv_fifo_elems,
-                                                 read_index_loc);
-
-    /* check the read_index to see if there is a new item to read (checking the owner bit) */
-    if (((read_index >> iface->fifo_shift) & 1) == ((read_index_elem->flags) & 1)) {
+    return (((iface->read_index >> iface->fifo_shift) & 1) ==
+            ((iface->read_index_elem->flags) & 1));
+}
 
+static inline unsigned
+uct_mm_iface_poll_fifo(uct_mm_iface_t *iface)
+{
+    /* check the read_index to see if there is a new item to read
+     * (checking the owner bit) */
+    if (uct_mm_iface_fifo_has_new_data(iface)) {
         /* read from read_index_elem */
         ucs_memory_cpu_load_fence();
         ucs_assert(iface->read_index <= iface->recv_fifo_ctl->head);
 
-        status = uct_mm_iface_process_recv(iface, read_index_elem);
-        if (status != UCS_OK) {
-            /* the last_recv_desc is in use. get a new descriptor for it */
-            UCT_TL_IFACE_GET_RX_DESC(&iface->super.super, &iface->recv_desc_mp,
-                                     iface->last_recv_desc, ucs_debug("recv mpool is empty"));
-        }
+        uct_mm_iface_process_recv(iface, iface->read_index_elem);
 
-        /* raise the read_index. */
+        /* raise the read_index */
         iface->read_index++;
 
+        /* the next fifo_element which the read_index points to */
+        iface->read_index_elem =
+            UCT_MM_IFACE_GET_FIFO_ELEM(iface, iface->recv_fifo_elems,
+                                       (iface->read_index & iface->fifo_mask));
+
         uct_mm_progress_fifo_tail(iface);
 
         return 1;
-    } else {
-        return 0;
     }
+
+    return 0;
 }
 
-unsigned uct_mm_iface_progress(void *arg)
+/*static inline void
+uct_mm_iface_fifo_window_update(uct_mm_iface_t *iface,
+                                )*/
+
+static unsigned uct_mm_iface_progress(uct_iface_h tl_iface)
 {
-    uct_mm_iface_t *iface = arg;
+    uct_mm_iface_t *iface = ucs_derived_of(tl_iface, uct_mm_iface_t);
+    unsigned total_count  = 0;
+    unsigned avg_count;
     unsigned count;
 
-    /* progress receive */
-    count = uct_mm_iface_poll_fifo(iface);
+    do {/* progress receive */
+        count        = uct_mm_iface_poll_fifo(iface);
+        ucs_assert((count == 0) || (count == 1));
+        total_count += count;
+        ucs_assert(total_count < UINT_MAX);
+    } while ((count != 0) && (total_count < iface->fifo_poll_window_size));
+
+    if (total_count != 0) {
+        if (++iface->fifo_window_update_iter < iface->config.fifo_window_update_max_iter) {
+            /* Calculate average value between the current FIFO window size
+             * and the total receive count for this iterationy */
+            avg_count = iface->fifo_poll_window_size + total_count;
+            avg_count = avg_count >> 1;
+
+            if (avg_count < iface->fifo_poll_window_size) {
```

Comment:
why can't just compare `total_count` and `iface->fifo_poll_window_size`?
imo, no need for average